### PR TITLE
Add SpinButton Component #261

### LINKED ISSUE #261: Add SpinButton Component
Add **SpinButton** Component based on [Microsoft FluentUI Design](https://developer.microsoft.com/en-us/fluentui#/controls/web/spinbutton).

### PR DESCRIPTION
**Pull request by HamedMoghadasi**
This closes #261

### CODE REVIEW

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                var result = Value - Step;
+                if (result >= Min)
+                {
+                    Value = Normalize(result);
+                    await OnChange.InvokeAsync(ValueWithSuffix);
+                }
+            }
+        }
+
+        protected override string RootElementClass => "bit-spin-button";
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs:58]: based on new conventions, it is better to choose a phrase like "bit-spb" or "bit-spn-btn"

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        protected override void RegisterComponentClasses()
+        {
+            ClassBuilder.Register(() => $"label-{LabelPosition}");
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs:62]: This implementation is kinda not correct.
Please take a look at other components (like BitButton) to grasp the correct implementation of this method.

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        protected virtual async Task HandleValueOnInputChange(ChangeEventArgs e)
+        {
+            if (IsEnabled)
+            {
+                var userInput = e.Value.ToString();
+                double numericValue;
+                var isNumber = double.TryParse(userInput, out numericValue);
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs:71]: make the declaration inline:
```cs
var isNumber = double.TryParse(userInput, out var numericValue);
```

File: src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/colors.fluent.scss [hunk 4]
```diff
@@ -26,7 +27,7 @@ $ThemeDarker: #004578;
...
 
 //Error & Status
-$Warning: #FFF4CE;
-$Error: #FDE7E9;
-$Success: #DFF6DD;
-$Severe: #FED9CC;
+$Warning: #fff4ce;
+$Error: #fde7e9;
+$Success: #dff6dd;
+$Severe: #fed9cc;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/colors.fluent.scss:33]: it is better to apply these changes on a separate PR based on a separate issue.

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/SpinButtons/bit-spin-button.fluent.scss [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    }
+
+    &.label-bottom {
+        flex-direction: column-reverse;
+    }
+}
+
+
+
+.bit-spb {
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/SpinButtons/bit-spin-button.fluent.scss:126]: Is it a duplicate code?
There is a ".bit-spb" at the top of this file.

File: src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/colors.fluent.scss [hunk 6]
```diff
@@ -4,15 +4,15 @@ $DarkAlt: #106EBE;
 $Dark: #005A9E;
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/colors.fluent.scss:4]: please undo the unnecessary changes in this file.

**Comment by msynk** [comment]
please remove ".css" files in this commit.

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor [hunk 7]
```diff
@@ -3,6 +3,35 @@
...
+
+    <h4># SpinButton with custom handler</h4>
+    <BitSpinButton DefaultValue="5" Min="0" Max="100" Step="1" Label="With custom handler" OnChange="@(e => HandleOnChange(e))"></BitSpinButton>
+    <p>@Message</p>
+
+    <h4># SpinButton with label above</h4>
+        <BitSpinButton DefaultValue="7" Suffix="cm" Min="0" Max="100" Step="1" Label="With labal above" LabelPosition="@LabelPosition.top"></BitSpinButton>
+</div>
+
+@code{
```
Comment (msynk) [inline src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor:27]: please move this "@code" section into the code behind file (Component.razor.cs)

**Review by msynk — CHANGES_REQUESTED** [review]

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        [Parameter] public string IconName { get; set; }
+        [Parameter] public LabelPosition LabelPosition { 
+            get => labelPosition; 
+            set {
+                labelPosition = value;
+                ClassBuilder.Reset();
+            }}
+        [Parameter] public EventCallback<string> OnChange { get; set; }
+
+        protected virtual async Task HandleUpClick(MouseEventArgs e)
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs:32]: why "protected virtual"?

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                var result = Value + Step;
+                if (result <= Max)
+                {
+                    Value = Normalize(result);
+                    await OnChange.InvokeAsync(ValueWithSuffix);
+                }
+            }
+        }
+
+        protected virtual async Task HandleDownClick(MouseEventArgs e)
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs:45]: why "protected virtual" here too?

**Comment by msynk** [comment]
because of inactivity for a long time in this PR, I'm closing this issue until further updates.

**Comment by msynk** [comment]
since the author came back, after a long hiatus I'm going to reopen this PR.

File: src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor (hunk 7 above)
Comment (HamedMoghadasi) [reply src/Client/Web/Playground/Bit.Client.Web.BlazorUI.Playground/Web/Pages/Component.razor]: done

File: src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/SpinButtons/bit-spin-button.fluent.scss (hunk 5 above)
Comment (HamedMoghadasi) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Styles/Fluent/SpinButtons/bit-spin-button.fluent.scss]: yes, it was a duplicate code.

File: src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        [Parameter] public double Min { get; set; } = 0;
+        [Parameter] public double Max { get; set; } = double.MaxValue;
+        [Parameter] public double DefaultValue { get; set; } = 0;
+        [Parameter] public double Step { get; set; } = 1;
+        [Parameter] public string Suffix { get; set; }
+        [Parameter] public string Label { get; set; }
+        [Parameter] public string IconName { get; set; }
+        [Parameter] public LabelPosition LabelPosition { 
+            get => labelPosition; 
+            set {
```
[the remark this answers is not in the document]
Comment (HamedMoghadasi) [reply src/Client/Web/Bit.Client.Web.BlazorUI/Components/SpinButtons/BitSpinButton.razor.cs]: done

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTest.razor [hunk 11]
```diff
@@ -0,0 +1,8 @@
+﻿<BitSpinButton Min="@Min" Max="@Max" Suffix="@Suffix" LabelPosition="@LabelPosition"></BitSpinButton>
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTest.razor:1]: please place each parameter in a separate line

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+using Microsoft.AspNetCore.Components;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Bit.Client.Web.BlazorUI.Tests.SpinButtons
+{
+    [TestClass]
+    public class BitSpinButtonTests : BunitTestContext
+    {
+        [TestInitialize]
+        public void SetuoMode()
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:16]: typo in the method name, and also please select a better name for this method

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            Context.JSInterop.Mode = JSRuntimeMode.Loose;
+        }
+
+        [DataTestMethod, DataRow("cm"), DataRow("Inch"), DataRow("foot")]
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:21]: please place each DataRow in a separate line

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            var component = RenderComponent<BitSpinButtonTest>(parameters => parameters.Add(p => p.Suffix, suffix));
+
+            var input = component.Find("input");
+            var inputValue = input.GetAttribute("value");
+
+            Assert.IsTrue(inputValue.Contains(suffix));
+        }
+
+        [DataTestMethod, DataRow(2, 4, 2), DataRow(20, 22, 20)]
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:32]: please place each DataRow in a separate line

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            for (int i = 0; i < countOfClick; i++)
+            {
+                upButton.Click();
+            }
+            var inputValue = double.Parse(input.GetAttribute("value"));
+
+            Assert.AreEqual(inputValue, expectedResult);
+        }
+
+        [DataTestMethod, DataRow(0, 4, 0), DataRow(-2, 22, -2)]
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:48]: please place each DataRow in a separate line

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            for (int i = 0; i < countOfClick; i++)
+            {
+                downButton.Click();
+            }
+            var inputValue = double.Parse(input.GetAttribute("value"));
+
+            Assert.AreEqual(inputValue, expectedResult);
+        }
+
+        [DataTestMethod, DataRow(LabelPosition.left), DataRow(LabelPosition.top)]
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:64]: please place each DataRow in a separate line

File: src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        public void SpinButtonShouldHaveLabelPositionClassName(LabelPosition labelPosition)
+        {
+            var component = RenderComponent<BitSpinButtonTest>(parameters => parameters.Add(p => p.LabelPosition, labelPosition));
+
+            var container = component.Find(".bit-spb");
+            var classlist = container.ClassList.ToString();
```
Comment (msynk) [inline src/Client/Web/Bit.Client.Web.BlazorUI.Tests/SpinButtons/BitSpinButtonTests.cs:70]: there is no need to convert the ClassList into a string

**Review by msynk — CHANGES_REQUESTED** [review]
please resolve the Nullable Reference Type warnings in files which have been changed in this PR

**Comment by msynk** [comment]
because this PR made from the develop branch, I'm closing it.
